### PR TITLE
Process ANSI colors in log viewer. Close #1958

### LINKED ISSUE #1958: Not converting red ANSI code in log viewer
DeepinScreenshot_select-area_20201013172015

### PR DESCRIPTION
**Pull request by umesh-timalsina**

### CODE REVIEW

File: src/visualizers/widgets/LogViewer/AnsiParser.js [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+        if (arr && Array.isArray(arr[0])) {
+            for (; i < arr.length; ++i) {
+                str += AnsiParser.stringify(arr[i]) + "\n";
+            }
+            str = str.replace(/\n$/, "");
+            return str;
+        }
+
+        var lastStyle = ""
```
Comment (brollb) [inline src/visualizers/widgets/LogViewer/AnsiParser.js:183]: Why did you change style conventions to use the leading comma?

File: src/visualizers/widgets/LogViewer/AnsiParser.js [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+                    , content: noAnsi[i++]
+                });
+                ++sIndex;
+            }
+        }
+
+        return result;
+    };
+
+    /**
```
Comment (brollb) [inline src/visualizers/widgets/LogViewer/AnsiParser.js:85]: Why do we have this function?

File: src/visualizers/widgets/LogViewer/AnsiParser.js [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            , content: c
+            , start: e || ""
+        });
+    };
+
+    return AnsiParser;
+}
+
+if (typeof define !== 'undefined') {
+    define([], () => factory);
```
Comment (brollb) [inline src/visualizers/widgets/LogViewer/AnsiParser.js:243]: I think this is meant to be
```javascript
define([], factory);
```

File: src/visualizers/widgets/LogViewer/LogViewerWidget.js [hunk 4]
```diff
@@ -17,6 +30,7 @@ define([
             lineNumbers: this.getLineNumbers
         });
         this.setReadOnly(true);
+        this.ansiParser = AnsiParser();
```
Comment (brollb) [inline src/visualizers/widgets/LogViewer/LogViewerWidget.js:33]: When the `define` is fixed above, this won't need to be invoked (technically, `AnsiParser` is the `factory` function).

File: src/visualizers/widgets/OperationCodeEditor/OperationCodeEditorWidget.js [hunk 5]
```diff
@@ -18,10 +18,10 @@ define([
         TextEditorWidget.call(this, logger, container);
         this.lineOffset = 0;
         // Add the shift-enter command
-        this.editor.addCommand(
-            monaco.KeyMod.Shift | monaco.KeyCode.Enter,
-            this.executeOrStopJob
-        );
+        // this.editor.addCommand(
```
Comment (brollb) [inline src/visualizers/widgets/OperationCodeEditor/OperationCodeEditorWidget.js:21]: Why did you remove this?

**Review by brollb — CHANGES_REQUESTED** [review]
I haven't tested it just yet but left some comments!

File: src/visualizers/widgets/LogViewer/AnsiParser.js (hunk 1 above)
Comment (umesh-timalsina) [reply src/visualizers/widgets/LogViewer/AnsiParser.js]: This module was used from a third-party library. It has been moved to `lib` directory.

File: src/visualizers/widgets/LogViewer/AnsiParser.js (hunk 2 above)
Comment (umesh-timalsina) [reply src/visualizers/widgets/LogViewer/AnsiParser.js]: Same comment as above.

File: src/visualizers/widgets/OperationCodeEditor/OperationCodeEditorWidget.js (hunk 5 above)
Comment (umesh-timalsina) [reply src/visualizers/widgets/OperationCodeEditor/OperationCodeEditorWidget.js]: There has been an additional check in the `addCommand` function where it checks if the handler is undefined (on update to new version of the monaco-editor). Since the widget gets initialized before control, I have changed it to be a function in the widget itself. Commit Reference [52272ed](https://github.com/deepforge-dev/deepforge/pull/2063/commits/52272edc05ecc0a289c38bd192651995e15e63dc).

File: src/visualizers/widgets/OperationCodeEditor/OperationCodeEditorWidget.js [hunk 6]
```diff
@@ -20,13 +20,17 @@ define([
...
-            this.executeOrStopJob
+            this.handleOperationExecute.bind(this)
         );
         this.editor.updateOptions({
             lineNumbers: this.updateOffset.bind(this)
         });
     };
 
+    OperationCodeEditorWidget.prototype.handleOperationExecute = function () {
+        return this.executeOrStopJob(arguments);
```
Comment (brollb) [inline src/visualizers/widgets/OperationCodeEditor/OperationCodeEditorWidget.js:31]: I believe this should be `...arguments` instead. The current approach will pass a single argument - the array-like arguments passed to `handleOperationExecute`.

**Review by brollb — APPROVED** [review]
Overall, it looks good. I left a minor comment or two that should be addressed then it should be good to merge!

File: src/visualizers/widgets/OperationCodeEditor/OperationCodeEditorWidget.js (hunk 6 above)
Comment (umesh-timalsina) [reply src/visualizers/widgets/OperationCodeEditor/OperationCodeEditorWidget.js]: This has been added.
